File: scripts/convert-sudoku.js
#!/usr/bin/env node

/**
 * Converts puzzles from the Sudoku Exchange Puzzle Bank into puzzle packs.
 *
 * Source: https://github.com/grantm/sudoku-exchange-puzzle-bank (Public Domain)
 * Puzzles are graded by Sukaku Explainer using technique-based difficulty ratings.
 *
 * Usage:
 *   node scripts/convert-sudoku.js <puzzle-bank-dir> [puzzles-per-difficulty] [pack-size]
 *
 * Example:
 *   node scripts/convert-sudoku.js /tmp/sudoku-bank 500 100
 *
 * Output:
 *   games/sudoku/puzzles/pack-001.js ... pack-NNN.js
 *   games/sudoku/puzzles/index.js
 */

import { readFileSync, writeFileSync, mkdirSync, readdirSync, unlinkSync } from 'fs';
import { resolve } from 'path';

const bankDir = process.argv[2];
const PER_DIFFICULTY = parseInt(process.argv[3], 10) || 500;
const PACK_SIZE = parseInt(process.argv[4], 10) || 100;

if (!bankDir) {
  console.error('Usage: node scripts/convert-sudoku.js <puzzle-bank-dir> [puzzles-per-difficulty] [pack-size]');
  process.exit(1);
}

const outDir = resolve(import.meta.dirname, '..', 'games', 'sudoku', 'puzzles');
mkdirSync(outDir, { recursive: true });

// ─── Step 1: Parse puzzle files ─────────────────────────────────────────────

const DIFFICULTY_MAP = {
  'easy.txt': 'easy',
  'medium.txt': 'medium',
  'hard.txt': 'hard',
  'diabolical.txt': 'expert',
};

/** Solve a sudoku puzzle using backtracking, return the solution grid */
function solve(grid) {
  const board = grid.map(row => [...row]);

  function isValid(row, col, num) {
    for (let i = 0; i < 9; i++) {
      if (board[row][i] === num) return false;
      if (board[i][col] === num) return false;
    }
    const br = Math.floor(row / 3) * 3;
    const bc = Math.floor(col / 3) * 3;
    for (let r = br; r < br + 3; r++) {
      for (let c = bc; c < bc + 3; c++) {
        if (board[r][c] === num) return false;
      }
    }
    return true;
  }

  function fill(pos) {
    if (pos === 81) return true;
    const r = Math.floor(pos / 9);
    const c = pos % 9;
    if (board[r][c] !== 0) return fill(pos + 1);
    for (let n = 1; n <= 9; n++) {
      if (isValid(r, c, n)) {
        board[r][c] = n;
        if (fill(pos + 1)) return true;
        board[r][c] = 0;
      }
    }
    return false;
  }

  if (!fill(0)) return null;
  return board;
}

/** Parse an 81-char digit string into a 9x9 grid */
function parseGrid(digits) {
  const grid = [];
  for (let r = 0; r < 9; r++) {
    const row = [];
    for (let c = 0; c < 9; c++) {
      row.push(parseInt(digits[r * 9 + c], 10));
    }
    grid.push(row);
  }
  return grid;
}

const allPuzzles = [];
let globalId = 1;

for (const [filename, diff] of Object.entries(DIFFICULTY_MAP)) {
  const filepath = resolve(bankDir, filename);
  console.log(`Reading ${filename}...`);

  const raw = readFileSync(filepath, 'utf-8');
  const lines = raw.trim().split('\n');

  // Shuffle lines to get a random sample
  for (let i = lines.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [lines[i], lines[j]] = [lines[j], lines[i]];
  }

  let count = 0;
  for (const line of lines) {
    if (count >= PER_DIFFICULTY) break;

    // Format: 12-byte hash, space, 81-byte digits, space, rating
    const parts = line.trim().split(/\s+/);
    if (parts.length < 3) continue;

    const digits = parts[1];
    const rating = parseFloat(parts[2]);
    if (digits.length !== 81 || !/^[0-9]+$/.test(digits)) continue;

    const puzzleGrid = parseGrid(digits);
    const solutionGrid = solve(puzzleGrid);
    if (!solutionGrid) {
      console.warn(`  Skipping unsolvable puzzle: ${digits.slice(0, 12)}...`);
      continue;
    }

    allPuzzles.push({
      id: globalId++,
      difficulty: diff,
      rating,
      puzzle: puzzleGrid,
      solution: solutionGrid,
    });

    count++;
    if (count % 100 === 0) {
      console.log(`  ${diff}: ${count}/${PER_DIFFICULTY} puzzles solved...`);
    }
  }

  console.log(`  ${diff}: ${count} puzzles converted`);
}

console.log(`\nTotal puzzles: ${allPuzzles.length}`);

// ─── Step 2: Write puzzle packs ─────────────────────────────────────────────

// Group by difficulty for the index, but pack sequentially
const totalPacks = Math.ceil(allPuzzles.length / PACK_SIZE);

const countByDifficulty = {};
for (const p of allPuzzles) {
  countByDifficulty[p.difficulty] = (countByDifficulty[p.difficulty] || 0) + 1;
}

for (let p = 0; p < totalPacks; p++) {
  const start = p * PACK_SIZE;
  const end = Math.min(start + PACK_SIZE, allPuzzles.length);
  const pack = allPuzzles.slice(start, end);
  const padded = String(p + 1).padStart(3, '0');

  const js = `export const PUZZLES = ${JSON.stringify(pack)};\n`;
  writeFileSync(`${outDir}/pack-${padded}.js`, js);
  console.log(`  pack-${padded}.js: ${pack.length} puzzles`);
}

// Remove stale pack files from previous runs
for (const file of readdirSync(outDir)) {
  if (/^pack-\d+\.js$/.test(file)) {
    const num = parseInt(file.match(/\d+/)[0], 10);
    if (num > totalPacks) {
      unlinkSync(`${outDir}/${file}`);
      console.log(`  Removed stale ${file}`);
    }
  }
}

// Write index
const indexJs = `export const TOTAL_PACKS = ${totalPacks};
export const PACK_SIZE = ${PACK_SIZE};
export const TOTAL_PUZZLES = ${allPuzzles.length};
export const PUZZLES_PER_DIFFICULTY = ${JSON.stringify(countByDifficulty)};
`;
writeFileSync(`${outDir}/index.js`, indexJs);

console.log(`\nDone! ${totalPacks} packs written to ${outDir}`);
console.log('Puzzles per difficulty:', countByDifficulty);

File: scripts/convert-sudoku.py
#!/usr/bin/env python3

"""
Converts puzzles from the Sudoku Exchange Puzzle Bank into puzzle packs.

Source: https://github.com/grantm/sudoku-exchange-puzzle-bank (Public Domain)
Puzzles are graded by Sukaku Explainer using technique-based difficulty ratings.

Usage:
  python scripts/convert-sudoku.py <puzzle-bank-dir> [puzzles-per-difficulty] [pack-size]

Example:
  python scripts/convert-sudoku.py /tmp/sudoku-bank 500 100

Output:
  games/sudoku/puzzles/pack-001.js ... pack-NNN.js
  games/sudoku/puzzles/index.js
"""

import json
import os
import random
import re
import sys

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
OUT_DIR = os.path.join(SCRIPT_DIR, '..', 'games', 'sudoku', 'puzzles')


def solve(grid):
    """Solve a sudoku puzzle using backtracking, return the solution grid or None."""
    board = [row[:] for row in grid]

    def is_valid(row, col, num):
        for i in range(9):
            if board[row][i] == num:
                return False
            if board[i][col] == num:
                return False
        br = (row // 3) * 3
        bc = (col // 3) * 3
        for r in range(br, br + 3):
            for c in range(bc, bc + 3):
                if board[r][c] == num:
                    return False
        return True

    def fill(pos):
        if pos == 81:
            return True
        r = pos // 9
        c = pos % 9
        if board[r][c] != 0:
            return fill(pos + 1)
        for n in range(1, 10):
            if is_valid(r, c, n):
                board[r][c] = n
                if fill(pos + 1):
                    return True
                board[r][c] = 0
        return False

    if not fill(0):
        return None
    return board


def parse_grid(digits):
    """Parse an 81-char digit string into a 9x9 grid."""
    grid = []
    for r in range(9):
        row = []
        for c in range(9):
            row.append(int(digits[r * 9 + c]))
        grid.append(row)
    return grid


def main():
    if len(sys.argv) < 2:
        print('Usage: python scripts/convert-sudoku.py <puzzle-bank-dir> [puzzles-per-difficulty] [pack-size]',
              file=sys.stderr)
        sys.exit(1)

    bank_dir = sys.argv[1]
    per_difficulty = int(sys.argv[2]) if len(sys.argv) > 2 else 500
    pack_size = int(sys.argv[3]) if len(sys.argv) > 3 else 100

    os.makedirs(OUT_DIR, exist_ok=True)

    # ─── Step 1: Parse puzzle files ─────────────────────────────────────────────

    difficulty_map = {
        'easy.txt': 'easy',
        'medium.txt': 'medium',
        'hard.txt': 'hard',
        'diabolical.txt': 'expert',
    }

    all_puzzles = []
    global_id = 1

    for filename, diff in difficulty_map.items():
        filepath = os.path.join(bank_dir, filename)
        print(f'Reading {filename}...')

        with open(filepath, 'r') as f:
            lines = f.read().strip().split('\n')

        # Shuffle lines to get a random sample
        random.shuffle(lines)

        count = 0
        for line in lines:
            if count >= per_difficulty:
                break

            # Format: 12-byte hash, space, 81-byte digits, space, rating
            parts = line.strip().split()
            if len(parts) < 3:
                continue

            digits = parts[1]
            try:
                rating = float(parts[2])
            except ValueError:
                continue

            if len(digits) != 81 or not re.match(r'^[0-9]+$', digits):
                continue

            puzzle_grid = parse_grid(digits)
            solution_grid = solve(puzzle_grid)
            if solution_grid is None:
                print(f'  Skipping unsolvable puzzle: {digits[:12]}...')
                continue

            all_puzzles.append({
                'id': global_id,
                'difficulty': diff,
                'rating': rating,
                'puzzle': puzzle_grid,
                'solution': solution_grid,
            })
            global_id += 1
            count += 1

            if count % 100 == 0:
                print(f'  {diff}: {count}/{per_difficulty} puzzles solved...')

        print(f'  {diff}: {count} puzzles converted')

    print(f'\nTotal puzzles: {len(all_puzzles)}')

    # ─── Step 2: Write puzzle packs ─────────────────────────────────────────────

    total_packs = (len(all_puzzles) + pack_size - 1) // pack_size

    count_by_difficulty = {}
    for p in all_puzzles:
        count_by_difficulty[p['difficulty']] = count_by_difficulty.get(p['difficulty'], 0) + 1

    for p in range(total_packs):
        start = p * pack_size
        end = min(start + pack_size, len(all_puzzles))
        pack = all_puzzles[start:end]
        padded = str(p + 1).zfill(3)

        # Write as ES6 module with compact JSON (no spaces after separators)
        js = f'export const PUZZLES = {json.dumps(pack, separators=(",", ":"))};\n'
        pack_path = os.path.join(OUT_DIR, f'pack-{padded}.js')
        with open(pack_path, 'w') as f:
            f.write(js)
        print(f'  pack-{padded}.js: {len(pack)} puzzles')

    # Remove stale pack files from previous runs
    for file in os.listdir(OUT_DIR):
        if re.match(r'^pack-\d+\.js$', file):
            num = int(re.search(r'\d+', file).group())
            if num > total_packs:
                os.unlink(os.path.join(OUT_DIR, file))
                print(f'  Removed stale {file}')

    # Write index
    index_js = (
        f'export const TOTAL_PACKS = {total_packs};\n'
        f'export const PACK_SIZE = {pack_size};\n'
        f'export const TOTAL_PUZZLES = {len(all_puzzles)};\n'
        f'export const PUZZLES_PER_DIFFICULTY = {json.dumps(count_by_difficulty, separators=(",", ":"))};\n'
    )
    with open(os.path.join(OUT_DIR, 'index.js'), 'w') as f:
        f.write(index_js)

    print(f'\nDone! {total_packs} packs written to {OUT_DIR}')
    print('Puzzles per difficulty:', count_by_difficulty)


if __name__ == '__main__':
    main()
